CORDA-3848 Uncaught exception hospitalises flow

lankydan opened the pull request:
When an uncaught exception propagates all the way to the flow exception 
handler, the flow will be forced into observation/hospitalised.

The updating of the checkpoints status is done on a separate thread as 
the fiber cannot be relied on anymore. The new thread is needed to allow
database transaction to be created and committed. Failures to the status
update will be rescheduled to ensure that this information is eventually
reflected in the database.

lankydan commented:
I think I will add some more tests to this but its a start.

williamvigorr3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -175,6 +175,8 @@ class SingleThreadedStateMachineManager(
             } else {
                 (fiber as FlowStateMachineImpl<*>).logger.warn("Caught exception from flow", throwable)
+                (fiber.resultFuture as? OpenFuture<*>)?.setException(throwable)
+                decrementLiveFibers()
```
Do we also need to decrement the unfinishedFibers?

kyriathar replied:
I think i agree with @williamvigorr3. This case here is different than when aborting a fiber due to an error. In that case the fiber could still be around, that's why we `while (true) { Fiber.park() }`. In this case the fiber object should be gone for sure i think. Hence we should also decrement the unfinishedFibers i think.

williamvigorr3 commented on core-tests/src/test/kotlin/net/corda/coretests/flows/FlowExternalAsyncOperationTest.kt:
```diff
@@ -137,18 +138,27 @@ class FlowExternalAsyncOperationTest : AbstractFlowExternalOperationTest() {
                 ).returnValue.getOrThrow(1.minutes)
             }
-            assertHospitalCounters(1, 0)
+            assertTrue("CordaRuntimeException or DirectlyAccessedServiceHubException should be thrown") {
```
Why does this test throw either `CordaRuntimeException` or `DirectlyAccessedServiceHubException` ?

lankydan replied:
couldnt diagnose the quasar issue, going to remove this test change anyway as it was to address an issue dimos had, which he isn't getting any more.

rick-r3 commented:
What's the scenario here?  An exception is usually caught by the try/catch in `FlowStateMachineImpl.run()` and presumably that ends up on the `Future` ?  Is this coping with something that then throws during the handling of that ?

lankydan commented:
> What's the scenario here? An exception is usually caught by the try/catch in FlowStateMachineImpl.run() and presumably that ends up on the Future ? Is this coping with something that then throws during the handling of that ?

This is for really odd scenarios, basically a catch all. I added this because Dimos was getting a quasar error on his branch, preventing the flow from following normal error handling. This prevents the test hanging etc.

lankydan committed:
CORDA-3848 Complete future in flow exception handler

When an exception propagates all the way through the flow to the flow
default uncaught exception handler, the flow should have it's future
completed with an exception.

The `liveFibers` should be decremented as well, since this flow has
technically finished since it exploded.

lankydan committed:
CORDA-3848 Uncaught exception hospitalises flow

When an uncaught exception propagates all the way to the flow exception
handler, the flow will be forced into observation/hospitalised.

The updating of the checkpoints status is done on a separate thread as
the fiber cannot be relied on anymore. The new thread is needed to allow
database transaction to be created and committed. Failures to the status
update will be rescheduled to ensure that this information is eventually
reflected in the database.

rick-r3 reviewed (commented):
Looks okay I think.  Needs a rebase

lankydan committed:
CORDA-3848 Fix detekt errors
